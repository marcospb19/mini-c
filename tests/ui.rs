//! This is a test module that creates snapshots, and might do
//! UI testing in the future.
//!
//! This shouldn't be called ui.rs hmmmmmmmm.
//!
//! Snapshot todo list:
//! - [X] Source code.
//! - [X] AST.
//! - [ ] LLVM IR.
//! - [ ] Execution stdout output (UI).
//! - [ ] (maybe) Execution final state (variables, functions).

use fs_err as fs;
use mini_c::test_utils;

macro_rules! snapshot_example_file {
    ($example_path:literal) => {
        let file_contents = &fs::read_to_string($example_path)
            .unwrap_or_else(|err| panic!("Failed to read file {}: {err}.", $example_path));
        // Source code
        insta::assert_display_snapshot!(file_contents);
        // AST
        insta::assert_ron_snapshot!(test_utils::generate_ast(file_contents));
    };
}

#[test]
fn example_break() {
    snapshot_example_file!("examples/break.c");
}
#[test]
fn example_coins() {
    snapshot_example_file!("examples/coins.c");
}
#[test]
fn example_conds() {
    snapshot_example_file!("examples/conds.c");
}
#[test]
fn example_easy() {
    snapshot_example_file!("examples/easy.c");
}
#[test]
fn example_fib() {
    snapshot_example_file!("examples/fib.c");
}
#[test]
fn example_file_1() {
    snapshot_example_file!("examples/file_1.c");
}
#[test]
fn example_file_2() {
    snapshot_example_file!("examples/file_2.c");
}
#[test]
fn example_if_pass() {
    snapshot_example_file!("examples/if-pass.c");
}
#[test]
fn example_operators() {
    snapshot_example_file!("examples/operators.c");
}
#[test]
fn example_queen() {
    snapshot_example_file!("examples/queen.c");
}
#[test]
fn example_var() {
    snapshot_example_file!("examples/var.c");
}
#[test]
fn example_var_loop() {
    snapshot_example_file!("examples/var_loop.c");
}
#[test]
fn example_while() {
    snapshot_example_file!("examples/while.c");
}
