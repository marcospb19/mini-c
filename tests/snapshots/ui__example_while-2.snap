---
source: tests/ui.rs
expression: test_mini_c(file_contents)
---
Program(
  preamble: true,
  declarations: [
    Function(FunctionDeclaration(Int, "main", [], Some(Scope([
      VariableDeclaration(Int, [
        Normal("J"),
        Normal("K"),
      ]),
      VariableDeclaration(Int, [
        Array("A", 100),
        Array("B", 100),
      ]),
      VariableDeclaration(Int, [
        Normal("i"),
      ]),
      VariableDeclaration(Int, [
        Normal("j"),
      ]),
    ], [
      For(ForStatement((Some(Assignment(AssignmentExpression(Normal("i"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("i")), Less, Value(Int(100))))), Some(Assignment(AssignmentExpression(Normal("i"), Binary(BinaryExpression(VariableReference(Normal("i")), Add, Value(Int(1)))))))), Scope([], [
        For(ForStatement((Some(Assignment(AssignmentExpression(Normal("j"), Value(Int(0))))), Some(Binary(BinaryExpression(VariableReference(Normal("j")), Less, Value(Int(100))))), Some(Assignment(AssignmentExpression(Normal("j"), Binary(BinaryExpression(VariableReference(Normal("j")), Add, Value(Int(1)))))))), Scope([], [
          Expression(Assignment(AssignmentExpression(Array("A", VariableReference(Normal("i"))), Binary(BinaryExpression(VariableReference(Normal("i")), Add, VariableReference(Normal("j"))))))),
        ]))),
      ]))),
      Expression(Assignment(AssignmentExpression(Normal("J"), Value(Int(0))))),
      While(WhileStatement(Some(Binary(BinaryExpression(VariableReference(Normal("J")), Less, Value(Int(100))))), Scope([], [
        Expression(Assignment(AssignmentExpression(Normal("K"), Value(Int(0))))),
        While(WhileStatement(Some(Binary(BinaryExpression(VariableReference(Normal("K")), Less, Value(Int(100))))), Scope([], [
          Expression(Assignment(AssignmentExpression(Normal("K"), Binary(BinaryExpression(VariableReference(Normal("K")), Add, Value(Int(1))))))),
        ]))),
        If(IfStatement(Binary(BinaryExpression(VariableReference(Array("B", VariableReference(Normal("J")))), Equals, VariableReference(Array("A", VariableReference(Normal("J")))))), Scope([], [
          Break,
        ]), None)),
        Expression(Assignment(AssignmentExpression(Normal("J"), Binary(BinaryExpression(VariableReference(Normal("J")), Add, Value(Int(1))))))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Array("B", Value(Int(0)))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Array("B", Value(Int(1)))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Array("B", Value(Int(2)))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Array("B", Value(Int(3)))),
      ]))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Array("B", Value(Int(4)))),
      ]))),
      Return(ReturnStatement(Some(Value(Int(0))))),
    ])))),
  ],
)
